### PR TITLE
[Metricbeat] Add Linux pressure metricset

### PR DESCRIPTION
**Pull request by b-deam**
## What does this PR do?

This PR aims to introduce an additional metricset that collects Pressure Stall Information (PSI) for CPU, Memory, and IO.  These are available from kernel >=4.20.

These metrics are incredibly useful for debugging system performance, and more can be read about these metrics below:

- https://www.kernel.org/doc/html/latest/accounting/psi.html
- https://lwn.net/Articles/759658/
- https://facebookmicrosites.github.io/psi/docs/overview.html

## Why is it important?

These are valuable OS metrics that are useful for both debugging and monitoring system performance.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [X] Tests need work.

## How to test this PR locally

- Pull down, run on a Linux host (`ubuntu2004` works from `Vagrantfile`) that:
  - Is on kernel >=4.20  
  - `/proc/pressure` is available on

## Related issues

- Closes https://github.com/elastic/beats/issues/27223

### CODE REVIEW

File: metricbeat/module/linux/pressure/_meta/fields.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  type: group
+  release: beta
+  description: >
+    Linux pressure stall information metrics for cpu, memory, and io
+  fields:
+    - name: pressure
+      type: group
+      description: Linux pressure stall information
+      fields:
+        - name: cpu.some.10
```
Comment (fearful-symmetry) [inline metricbeat/module/linux/pressure/_meta/fields.yml:11]: In `system/process` cgroup data, we're reporting these as `10.pct` and so on. We'll also need to add `format: percent`

File: metricbeat/module/linux/pressure/pressure.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta(fmt.Sprintf("The %s %s metricset is beta.", moduleName, metricsetName))
+
+	if runtime.GOOS != "linux" {
+		return nil, fmt.Errorf("the %v/%v metricset is only supported on Linux", moduleName, metricsetName)
+	}
+
+	fs, err := procfs.NewFS("/proc")
```
Comment (fearful-symmetry) [inline metricbeat/module/linux/pressure/pressure.go:63]: We have a global `Path.Hostfs` that's used in other places, and we'll probably want to use it here, as it allows users to specify a custom root, for instances where they're running MB inside a container.

File: metricbeat/module/linux/pressure/pressure.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+	for _, event := range events {
+		report.Event(mb.Event{
+			MetricSetFields: event,
+		})
+	}
+	return nil
+}
+
+func FetchLinuxPSIStats(m *MetricSet) ([]common.MapStr, error) {
```
Comment (fearful-symmetry) [inline metricbeat/module/linux/pressure/pressure.go:91]: I don't think this needs to be exported?

File: metricbeat/module/linux/pressure/pressure_test.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	mbtest "github.com/elastic/beats/v7/metricbeat/mb/testing"
+)
+
+func TestFetch(t *testing.T) {
```
Comment (fearful-symmetry) [inline metricbeat/module/linux/pressure/pressure_test.go:30]: A mocked procfs root will also be useful here, as we can just specify a testdata file.

**Review by fearful-symmetry — COMMENTED** [review]
This is looking good! Just a few small things, mostly beats cruft, we need to change. If you want, I can take the PR over from here; I don't want to make you feel like you're obligated to spend your time coding to the various metricbeat idioms.

File: metricbeat/module/linux/pressure/_meta/fields.yml (hunk 1 above)
Comment (b-deam) [reply metricbeat/module/linux/pressure/_meta/fields.yml:11]: Cool. I tried to follow similar conventions for `total`, in that it becomes `total.time.us` - does this make sense here?

File: metricbeat/module/linux/pressure/pressure.go (hunk 2 above)
Comment (b-deam) [reply metricbeat/module/linux/pressure/pressure.go:63]: Yup! I've copied the implementation from the other modules.

File: metricbeat/module/linux/pressure/pressure.go (hunk 3 above)
Comment (b-deam) [reply metricbeat/module/linux/pressure/pressure.go:91]: You're right! Fixed.

File: metricbeat/module/linux/pressure/pressure_test.go (hunk 4 above)
Comment (b-deam) [reply metricbeat/module/linux/pressure/pressure_test.go:32]: Great idea. Thanks for the tip!

**Comment by b-deam** [comment]
>  If you want, I can take the PR over from here; I don't want to make you feel like you're obligated to spend your time coding to the various metricbeat idioms.

Not a problem - I _think_ I've addressed the comments so far but feel free to make any small changes if I've otherwise messed things up there.

**Comment by fearful-symmetry** [comment]
Real quick, you'll need to run `mage fmt update` in the main metricbeat directory.

**Comment by b-deam** [comment]
> Real quick, you'll need to run `mage fmt update` in the main metricbeat directory.

Hmm.. I tried this but I don't see any changes. Any ideas?

```
bradleydeam in ~/go/src/github.com/b-deam/beats/metricbeat on branch linux-pressure-metricset $ make update  
mage update
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/fields.yml
>> Building metricbeat.yml for linux/amd64
>> Building metricbeat.reference.yml for linux/amd64
>> Building metricbeat.docker.yml for linux/amd64
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/build/fields/fields.all.yml
exec: go list -m
exec: go list -m              

bradleydeam in ~/go/src/github.com/b-deam/beats/metricbeat on branch linux-pressure-metricset $ mage fmt update
>> fmt - go-licenser: Adding missing headers
>> fmt - goimports: Formatting Go code
>> fmt - autopep8: Formatting Python code
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/fields.yml
>> Building metricbeat.yml for linux/amd64
>> Building metricbeat.reference.yml for linux/amd64
>> Building metricbeat.docker.yml for linux/amd64
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/build/fields/fields.all.yml
exec: go list -m
exec: go list -m

bradleydeam in ~/go/src/github.com/b-deam/beats/metricbeat on branch linux-pressure-metricset $ git status
On branch linux-pressure-metricset
Your branch is up to date with 'origin/linux-pressure-metricset'.

nothing to commit, working tree clean
```

**Comment by fearful-symmetry** [comment]
@b-deam I think you'll need to re-merge upstream changes? A few things may have changed in the doc build.

**Comment by b-deam** [comment]
> @b-deam I think you'll need to re-merge upstream changes? A few things may have changed in the doc build.

🤦 - ofc. Done. 🤞 tests are good!

File: metricbeat/module/linux/pressure/pressure.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	if runtime.GOOS != "linux" {
+		return nil, fmt.Errorf("the %v/%v metricset is only supported on Linux", moduleName, metricsetName)
+	}
+
+	linuxModule, ok := base.Module().(*linux.Module)
+	if !ok {
+		return nil, errors.New("unexpected module type")
+	}
+
+	path := filepath.Join(linuxModule.HostFS, "proc")
```
Comment (fearful-symmetry) [inline metricbeat/module/linux/pressure/pressure.go:71]: I...didn't even know that was there. There's actually a `"github.com/elastic/beats/v7/libbeat/paths"` that's global and used by most things, but I guess linux has its own hostfs path. In the future we might want to change it, but for now we should probably keep linux internally consistent.

File: metricbeat/module/linux/pressure/pressure_test.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	assert.Empty(t, errs)
+	if !assert.NotEmpty(t, events) {
+		t.FailNow()
+	}
+	t.Logf("%s/%s event: %+v", f.Module().Name(), f.Name(),
+		events[0].BeatEvent("linux", "pressure").Fields.StringToPrint())
+
+	resources := []string{"cpu", "memory", "io"}
+
+	for i, _ := range events {
```
Comment (fearful-symmetry) [inline metricbeat/module/linux/pressure/pressure_test.go:45]: Nit: This can be simplified to `for i := range events`

**Review by fearful-symmetry — APPROVED** [review]
LGTM.

**Comment by b-deam** [comment]
> LGTM.

I re-ran a test to make sure it all looked good and noticed that I'd messed up the fields/mapping so that the metrics were actually recorded as:

```
linux.pressure.pressure.*
```

This should be fixed now..

**Comment by b-deam** [comment]
Thanks for all the help @fearful-symmetry!

**Pull request merged**